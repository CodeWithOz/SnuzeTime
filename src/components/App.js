import React, { Component } from 'react';
import Navbar from './Navbar';
import Clock from './Clock';
import ButtonDisplay from './ButtonDisplay';
import TodayView from './TodayView';
import moment from 'moment';
import 'moment-timer';

function getCurrentTime(withSeconds = true) {
  return moment().format(`hh:mm${withSeconds ? ':ss' : ''} A`);
}

class App extends Component {
  constructor(props) {
    super(props);

    this.state = {
      // TODO: improve the initial loading experience
      // currently, the time is always initialized at midnight
      currentTime: '00:00:00 AM',
      currentHour: 0,
      wakeTime: '',
      getUpTime: '',
      sleepTime: ''
    };

    this.startTimer();
  }

  startTimer = () => {
    new moment.duration(1000).timer({ start: true, loop: true }, () => {
      this.setState({
        currentTime: getCurrentTime(),
        currentHour: Number(moment().format('HH'))
      });
    });
  };

  saveSleepTime = () => {
    this.setState({ sleepTime: getCurrentTime(false) });
  };

  saveWakeTime = () => {
    this.setState({ wakeTime: getCurrentTime(false) });
  };

  saveGetUpTime = () => {
    this.setState({ getUpTime: getCurrentTime(false) });
  };

  render() {
    return (
      <section>
        <Navbar title="SnuzeTime" />
        <Clock currentTime={this.state.currentTime} />
        <ButtonDisplay
          hour={this.state.currentHour}
          saveSleepTime={this.saveSleepTime}
          saveWakeTime={this.saveWakeTime}
          saveGetUpTime={this.saveGetUpTime}
        />
        <hr />
        <TodayView
          sleepTime={this.state.sleepTime}
          wakeTime={this.state.wakeTime}
          getUpTime={this.state.getUpTime}
        />
      </section>
    );
  }
}

export default App;
