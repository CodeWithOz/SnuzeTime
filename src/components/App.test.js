import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import 'moment-timer';
import { App } from './App';

// mock native timers
jest.useFakeTimers();

test('App sets a 1-second recurring timer', () => {
  // clear previous calls to the mocked timers
  setInterval.mockClear();
  expect(setInterval).not.toHaveBeenCalled();

  // initiate the component
  shallow(<App />);

  expect(setInterval).toHaveBeenCalledTimes(1);
  expect(setInterval).toHaveBeenCalledWith(expect.any(Function), 1000);
});

describe('App gets the current', () => {
  test('time as HH:MM AM/PM', () => {
    const timeWithoutSeconds = App.prototype.getCurrentTime(false);

    // moment returns nicely formatted time strings
    const timeFromMoment = moment().format(`hh:mm A`);
    expect(timeWithoutSeconds).toEqual(timeFromMoment);
  });

  test('time as HH:MM:SS AM/PM', () => {
    const timeWithSeconds = App.prototype.getCurrentTime();

    const timeFromMoment = moment().format(`hh:mm:ss A`);
    expect(timeWithSeconds).toEqual(timeFromMoment);
  });

  test('hour in 24-hour format', () => {
    const hour = App.prototype.getCurrentHour();

    const timeFromMoment = Number(moment().format(`HH`));
    expect(hour).toEqual(timeFromMoment);
  });

  test('date as YYYY M D', () => {
    const date = App.prototype.getCurrentDate();

    const timeFromMoment = moment().format(`YYYY M D`);
    expect(date).toEqual(timeFromMoment);
  });
});

test('App gets the correct background color', () => {
  expect(App.prototype.getBackground).toBeDefined();

  const dayBkrgnd = 'light-1';
  const nightBkgrnd = 'dark-1';
  expect(App.prototype.getBackground(10)).toEqual(dayBkrgnd);
  expect(App.prototype.getBackground(22)).toEqual(nightBkgrnd);
});

test('App gets the correct spinner color', () => {
  expect(App.prototype.getSpinnerColor).toBeDefined();

  const nightColor = '#F8F8F8';
  const dayColor = '#333333';
  expect(App.prototype.getSpinnerColor(10)).toEqual(dayColor);
  expect(App.prototype.getSpinnerColor(22)).toEqual(nightColor);
});
