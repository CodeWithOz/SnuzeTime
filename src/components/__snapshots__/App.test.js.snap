// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
<div>
  <Clock
    currentTime="02:49 AM"
  />
</div>
`;
